### PR TITLE
[RISCV] Set DebugLoc of epilogue

### PR DESCRIPTION
**Pull request by wangpc-pp**
There is a problem I found when discussing at
https://discourse.llvm.org/t/llvm-generates-wrong-dwarf-line-table-to-gdb/75454/10:
if we set a breakpoint to the end of a function, the debugger will
stop at the place after restoring registers, not before it, which
is a wrong behavior I think.

There are two problems here actually:

1. Instructions for restoring registers don't have debug line info
  as we just set it to `DebugLoc()`.
2. We can't recognize the right epilogue beginning `epilogue_begin`.
  This is a feature introduced by https://reviews.llvm.org/D133376.
  In short, the epilogue beginning will be the first instruction
  with flag `FrameDestroy`. The problem for RISCV target is that
  we only set `FrameDestroy` flag for stack pointer recovering
  instructions(IIUC).

This PR fixes the first problem and the fix is to use the `DebugLoc`
getting from the insert point `I` like other in-tree targets.

As for second problem, I don't have a fix now and I think it may
be hard to fix, so I will leave it there.

A test is added to show the change, please see the topmost commit.

### CODE REVIEW

File: llvm/lib/Target/RISCV/RISCVInstrInfo.cpp [hunk 1]
```diff
@@ -681,6 +681,10 @@ void RISCVInstrInfo::loadRegFromStackSlot(MachineBasicBlock &MBB,
                                           const TargetRegisterClass *RC,
                                           const TargetRegisterInfo *TRI,
                                           Register VReg) const {
+  DebugLoc DL;
+  if (I != MBB.end())
```
[the remark this answers is not in the document]
Comment (wangpc-pp) [reply llvm/lib/Target/RISCV/RISCVInstrInfo.cpp]: Thanks! I use `MachineBasicBlock::findDebugLoc` now.

**Comment by topperc** [comment]
What do X86 or Aarch64 do? I'm not sure if I trust copying something from MIPS. MIPS doesn't seem to be getting many contributions these days.

**Comment by wangpc-pp** [comment]
> What do X86 or Aarch64 do? I'm not sure if I trust copying something from MIPS. MIPS doesn't seem to be getting many contributions these days.

Sorry for unclear description, most in-tree targets will use the `DebugLoc` getting from the insert point `I`. But the two targets (AArch64/X86) you mentioned don't do it in `loadRegFromStackSlot`... I don't know if they have the issue.

**Comment by asb** [comment]
I believe these were intentionally removed in https://github.com/llvm/llvm-project/commit/700042cd8890 which in turn references  3e081703c349 by @MatzeB.

You might also be interested in https://reviews.llvm.org/D143248 which sadly got completely stuck as nobody seems to be able to review it.

**Comment by asb** [comment]
> But is this the right way for current situation? The debugger behaves unproperly.

My standard assumption would be that following what X86/AArch64/Arm chose to do is correct and that remaining issues are likely caused by something else. I could be wrong of course, but I think you'd need to try to convince us as reviewers of why a different choice vs those targets is desirable here.

I really don't think there's been all that much critical analysis of the debuggability of LLVM generated RISC-V executables (though issues have been raised and fixed over the years), so I'm sure there are problems to be resolved like the one you raised in this PR.

**Comment by LiqinWeng** [comment]
pls see: 
1. https://reviews.llvm.org/D128806 
2. https://reviews.llvm.org/D129173
I remove it before
